third-party/rsutils/include/rsutils/signal.h

```diff
@@ -0,0 +1,150 @@
+// License: Apache 2.0. See LICENSE file in root directory.
+// Copyright(c) 2023 Intel Corporation. All Rights Reserved.
+
+#pragma once
+
+#include "subscription.h"
+
+#include <map>
+#include <mutex>
+#include <functional>
+#include <memory>
+
+
+namespace rsutils {
+
+
+// Outside signal<> so can be easy to reference - they should be the same regardless of the type of signal
+using subscription_slot = int;
+
+
+// Signals are callbacks with multiple targets.
+// 
+// Each "slot" is a subscriber that's called when the signal is "raised."
+// 
+// There are intentionally no operators here (especially operator()!) as, despite the nice-to-look-at style, are not
+// verbose enough to be really useful.
+// 
+// The template Args are the callback arguments, and therefore what's expected to be passed to raise(). The callbacks
+// return void on purpose, for simplicity.
+// 
+// NOTE: there are more efficient implementations (e.g., with a vector). We kept it map-based for simplicity.
+//
+template< typename... Args >
+class signal
+{
+    using callback = std::function< void( Args... ) >;
+    using map = std::map< subscription_slot, callback >;
+
+    // We use a shared_ptr to control lifetime: only while it's alive can we remove subscriptions
+    struct impl
+    {
+        std::mutex mutex;
+        map subscribers;
+
+        bool remove( subscription_slot slot )
+        {
+            std::lock_guard< std::mutex > locker( mutex );
+            return subscribers.erase( slot );
+        }
+    };
+    std::shared_ptr< impl > const _impl;
+
+    signal( const signal & other ) = delete;
+    signal & operator=( const signal & ) = delete;
+
+public:
+    signal() : _impl( std::make_shared< impl >() ) {}
+    signal( signal && ) = default;
+    signal & operator=( signal && ) = default;
+
+    // Get the slot without any fancy subscription
+    subscription_slot add( callback && func )
+    {
+        std::lock_guard< std::mutex > locker( _impl->mutex );
+        // NOTE: we should maintain ordering of subscribers: later subscriptions should be called after earlier ones, so
```

Comment:
Why do we want to keep order of calls? We should not rely on it when using a general utility.
We might not be able to guarantee this if moving to other implementations, e.g. boost.

Reply:
Because it's general-purpose is exactly why I wanted to keep the order.
If the order is non-deterministic, then the function (result of using all the callbacks) isn't deterministic. E.g., the unit-test actually has a callback that gets an `std::string &` argument -- it actually changes the output value. The caller decides the functionality of the callbacks, and therefore losing determinism is not a good thing IMO. It doesn't cost much, either.

Reply:
> We might not be able to guarantee this if moving to other implementations, e.g. boost.

That's true. I actually implemented using `vector<>`. It was faster to create and run the signals. And you're right, it would have been harder to make it deterministic, though possible. The problem was the complexity. In the end I chose to go back to a `map<>` implementation because of its straightforwardness. It's good enough for our needs.

We're not likely to change the implementation. And if we do, it's better if we ensure determinism then, too.